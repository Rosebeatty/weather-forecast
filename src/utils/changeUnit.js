export const changeUnit = (unit, setUnit) => {
  setUnit(!unit);
  localStorage.setItem("toggle", JSON.stringify(unit));
};
