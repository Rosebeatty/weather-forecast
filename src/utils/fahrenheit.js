export const fahrenheit = (c) => {
  return Math.round((c * 9) / 5 + 32);
};
