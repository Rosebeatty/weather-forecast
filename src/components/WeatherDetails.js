import styled from "styled-components";
import { fahrenheit } from "utils/fahrenheit";
import { format } from "date-fns";

const Grid = styled.div`
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  color: purple;
  margin: 0 auto;
  max-width: 75vw;
  @media only screen and (max-width: 720px) {
  }
`;
const Box = styled.div`
  padding: 0.7em;
  width: 25vw;

  &.box-1 {
    border-right: 1px solid purple;
    border-bottom: 1px solid purple;
  }
  &.box-2 {
    border-bottom: 1px solid purple;
  }
  &.box-4 {
    border-left: 1px solid purple;
  }
`;
const Card = styled.div`
  min-height: 35vh;
  max-width: 60vw;
  display: flex;
  color: purple;
  flex-direction: column;
  justify-content: center;
  margin: 1.7em auto;
  text-align: left;
  align-items: center;
  background-color: rgba(245, 245, 245, 0.2);
  border-radius: 5px;

  @media only screen and (max-width: 720px) {
    min-height: 38vh;
    min-width: 75vw;
  }
`;
const MaxMin = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  gap: 0.5em;
  width: 40vw;
`;
const Temp = styled.h1`
  font-size: 3.7em;
  margin: 0.2em 0;
`;

// Weather details display for chosen location
function WeatherDetails({ destination, selectedResult, unit, isLocal }) {
  const { current, current_units, daily, daily_units } = destination;
  const sunrise = format(new Date(daily.sunrise[0]), "HH:mm");
  const sunset = format(new Date(daily.sunset[0]), "HH:mm");

  return (
    <div>
      <h1>
        {(selectedResult &&
          `${selectedResult?.name}, ${selectedResult?.country}`) ||
          (isLocal === true && "Local Weather") ||
          `${destination.longitude}° long - ${destination.longitude}° lat`}
      </h1>
      <h2 style={{ fontSize: "16px" }}>
        {format(new Date(daily.time[0]), "EEEE LLLL do y")}
      </h2>
      <Card>
        <Temp>
          <p id="unit" data-testid="unit">
            {unit
              ? `${Math.round(current.temperature_2m)} ${current_units.temperature_2m}`
              : `${fahrenheit(current.temperature_2m)}°F`}
          </p>
        </Temp>
        <MaxMin>
          <p>
            Min:{" "}
            {unit
              ? `${Math.round(daily.temperature_2m_min)} ${daily_units.temperature_2m_min}`
              : `${fahrenheit(daily.temperature_2m_min)}°F`}
          </p>
          <p>
            Max:{" "}
            {unit
              ? `${Math.round(daily.temperature_2m_max)} ${daily_units.temperature_2m_max}`
              : `${fahrenheit(daily.temperature_2m_max)}°F`}
          </p>
        </MaxMin>
      </Card>
      <Grid>
        {selectedResult && (
          <>
            <Box className="box-1">
              <h4>Sunrise</h4>
              <p>{sunrise}</p>
            </Box>
            <Box className="box-2">
              <h4>Sunset</h4>
              <p>{sunset}</p>
            </Box>
          </>
        )}
        <Box className="box-1">
          <h4>Wind Speed</h4>
          <p>
            {current.wind_speed_10m} {current_units.wind_speed_10m}
          </p>
        </Box>
        <Box className="box-2">
          <h4>Percipitation</h4>
          <p>
            {current.precipitation} {current_units.precipitation}
          </p>
        </Box>
        <Box>
          <h4>Humidity</h4>
          <p>
            {current.relative_humidity_2m} {current_units.relative_humidity_2m}
          </p>
        </Box>
      </Grid>
    </div>
  );
}

export default WeatherDetails;
