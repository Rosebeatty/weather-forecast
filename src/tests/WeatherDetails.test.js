import React from "react";
import { render, screen } from "@testing-library/react";
import WeatherDetails from "components/WeatherDetails";
import { mockLocations } from "./mockData";
import { fahrenheit } from "utils/fahrenheit";
import { format } from "date-fns";

beforeEach(() => {
  const mockProps = {
    destination: mockLocations[0],
    unit: true,
  };
  render(<WeatherDetails {...mockProps} />);
});

it("renders without crashing", () => {
  const { asFragment } = render(
    <WeatherDetails destination={mockLocations[0]} unit={false} />,
  );
  expect(asFragment()).toMatchSnapshot();
});

it("rounds decimals to whole numbers", () => {
  const unitElement = screen.getByTestId("unit");
  expect(unitElement.textContent).not.toContain(".");
});

it("converts sunrise to datetime", () => {
  const sunrise = mockLocations[0].daily.sunrise[0];
  const date = format(new Date(sunrise), "HH:mm");
  expect(date).toEqual("05:41");
});

it("converts sunset to datetime", () => {
  const sunset = mockLocations[0].daily.sunset[0];
  const date = format(new Date(sunset), "HH:mm");
  expect(date).toEqual("21:15");
});
