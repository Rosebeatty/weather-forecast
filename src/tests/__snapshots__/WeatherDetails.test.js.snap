// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div>
    <h1>
      9.18° long - 9.18° lat
    </h1>
    <h2
      style="font-size: 16px;"
    >
      Monday July 8th 2024
    </h2>
    <div
      class="sc-eqUzNf jjSnNo"
    >
      <h1
        class="sc-dcJtft hprDHP"
      >
        <p
          data-testid="unit"
          id="unit"
        >
          58°F
        </p>
      </h1>
      <div
        class="sc-fqkwJk ldiSSK"
      >
        <p>
          Min: 58°F
        </p>
        <p>
          Max: 73°F
        </p>
      </div>
    </div>
    <div
      class="sc-aYaIB bHPGlB"
    >
      <div
        class="sc-gEvDqW eLaTiZ box-1"
      >
        <h4>
          Wind Speed
        </h4>
        <p>
          4.8 km/h
        </p>
      </div>
      <div
        class="sc-gEvDqW eLaTiZ box-2"
      >
        <h4>
          Precipitation
        </h4>
        <p>
          0.2 mm
        </p>
      </div>
      <div
        class="sc-gEvDqW eLaTiZ"
      >
        <h4>
          Humidity
        </h4>
        <p>
          99 %
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
